vignettes/mmrm_review_methods.Rmd

```diff
@@ -0,0 +1,582 @@
+mb <- microbenchmark(times = n_reps,
+  gls = gls(
+    FEV1 ~ RACE + ARMCD * AVISIT, data = fev_data,
+    na.action = na.omit,
+    correlation = nlme::corSymm(form = ~ VISITN | USUBJID),
+    weights = nlme::varIdent(form = ~ 1|VISITN)
+  ),
+  mmrm = mmrm(
+    formula = FEV1 ~ RACE + ARMCD * AVISIT + us(AVISIT | USUBJID),
+    data = fev_data
+  ),
+  lmer = lmer(
+    FEV1 ~ RACE + ARMCD * AVISIT + (0 + AVISIT | USUBJID),
+    data = fev_data,
+    control = lmerControl(check.nobs.vs.nRE = "ignore"),
+    na.action = na.omit
+  ),
+  glmmTMB = glmmTMB(
+    FEV1 ~ RACE + ARMCD * AVISIT + us(0 + AVISIT | USUBJID),
+    data = fev_data,
+    dispformula = ~ 0,
+    REML = TRUE
+  )
+ )
+
+# construct the partial results table based on R-based MMRM procedures
+partial_conv_time_tbl <- mb %>%
+  summary() %>%
+  arrange(median) %>%
+  select(expression = expr, median, lower = lq, upper = uq, evaluations = neval)
+
+# fit PROC MIXED 10 times, extract convergence times
+fit_times <- sapply(
+  seq_len(n_reps),
+  function(x) {
+    # load the data into SAS
+    df2sd(fev_data, "fev_data")
+
+    sas_code <- "
+      PROC MIXED DATA = fev_data cl method=reml;
+        CLASS AVISIT(ref = 'VIS1') RACE(ref = 'Asian') ARMCD(ref = 'PBO') USUBJID;
+        MODEL FEV1 = RACE ARMCD AVISIT ARMCD*AVISIT / ddfm=satterthwaite solution chisq;
+        REPEATED AVISIT / subject=USUBJID type=un r rcorr;
+     RUN;
+    "
+
+    # generate the SAS output
+    sas_result <- run_sas(sas_code)
+
+    # extract the fit time (seconds)
+    fit_time <- sas_result$LOG %>%
+      str_extract("(?<=real time)\\s*[0-9.]+") %>%
+      as.numeric()
+
+    # convert fit times from seconds to milliseconds
+    fit_time * 1000
+  }
+)
+
+# add PROC MIXED results to convergence results table
+proc_mixed_row <- tibble(
+  expression = "PROC MIXED",
+  median = median(fit_times),
+  lower = quantile(fit_times, probs = 0.25),
+  upper = quantile(fit_times, probs = 0.75),
+  evaluations = n_reps
+)
+conv_time_tbl <- bind_rows(partial_conv_time_tbl, proc_mixed_row)
+
+# format table in markdown
+conv_time_tbl %>%
+  knitr::kable(caption = "Comparison of convergence times: milliseconds", digits = 2)
+```
+
+The `mmrm`, `PROC MIXED`, `gls`, `lmer`, and `glmmTMB` functions are applied to
+the FEV dataset 100 times. The convergence times are recorded for each replicate
+and are reported in the table below.
+
+Table: Comparison of convergence times (milliseconds)
+
+| Implementation | Median | Lower Quartile | Upper Quartile |
+|:---------------|-------:|---------------:|---------------:|
+| `PROC MIXED`   |  20.00 |          20.00 |          20.00 |
+| `mmrm`         |  50.80 |          50.48 |          51.45 |
+| `lmer`         |  97.35 |          96.81 |          97.83 |
+| `glmmTMB`      | 273.45 |         271.77 |         275.28 |
+| `gls`          | 342.08 |         340.57 |         344.04 |
+
+It is clear from these results that `mmrm` converges significantly faster than
```

Comment:
Would it be possible to generate this table based on the actual results data frame instead of hardcoding it?

Also I Wonder of it would be good to add a much larger dataset as second example for these timings since there SAS might be much slower.

Reply:
Of course, it shouldn't be an issue!

That sounds good. Please see my comment below and let me know what you think.

Reply:
I added a dataset to my most recent pushes. Let me know what you think!